### PR TITLE
Read relational expression values from Json Path.

### PR DESCRIPTION
**Pull request by ravitejakula**
Input: Valid Jsonpath.
Output: List of SimpleEntry objects containing the left and right side values of the relational expression if any in the json path.

### CODE REVIEW

File: json-path/src/main/java/com/jayway/jsonpath/ReadContext.java [hunk 1]
```diff
@@ -14,6 +14,11 @@
  */
 package com.jayway.jsonpath;
 
+import java.util.AbstractMap.SimpleEntry;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
```
Comment (krdev) [inline json-path/src/main/java/com/jayway/jsonpath/ReadContext.java]: remove this, you don't have to change this file?? Make sure you have jdoc to interfaces, abstract classes and the public class (JsonPath)

File: json-path/src/main/java/com/jayway/jsonpath/internal/JsonContext.java [hunk 2]
```diff
@@ -143,9 +146,13 @@ public Object readRoot(String[] paths, Predicate... filters) {
             	}
             }
         }
+        if (ret == null) {
+        	// Return an empty map.
+        	ret = configuration.jsonProvider().createMap();
+        }
```
Comment (krdev) [inline json-path/src/main/java/com/jayway/jsonpath/internal/JsonContext.java:149]: Why do you need this?

File: json-path/src/main/java/com/jayway/jsonpath/Predicate.java [hunk 3]
```diff
@@ -22,6 +25,7 @@
 public interface Predicate {
 
     boolean apply(PredicateContext ctx);
+    void getRelationalExprValues(ArrayList<SimpleEntry<String,String>> valuesMap);
```
Comment (krdev) [inline json-path/src/main/java/com/jayway/jsonpath/Predicate.java]: Can you add jdoc for the interface

File: json-path/src/main/java/com/jayway/jsonpath/internal/JsonContext.java [hunk 4]
```diff
@@ -19,8 +19,11 @@
 import static com.jayway.jsonpath.internal.Utils.notNull;
 import static java.util.Arrays.asList;
 
+import java.util.ArrayList;
```
Comment (krdev) [inline json-path/src/main/java/com/jayway/jsonpath/internal/JsonContext.java]: Why do you need this? you are not changing this class??

File: json-path/src/main/java/com/jayway/jsonpath/internal/path/CompiledPath.java [hunk 5]
```diff
@@ -142,4 +147,17 @@ public boolean isFunctionPath() {
...
+    /**
+     * Fetches the relational expression values from this path.
+     *
+     * @return EvaluationContext containing results of evaluation
+     */
+    @SuppressWarnings("unchecked")
+	@Override
+    public ArrayList<SimpleEntry<String, String>> getRelationalExprValues(){
+    	ArrayList<SimpleEntry<String, String>> valuesMap = new ArrayList<SimpleEntry<String, String>>();
+    	root.next().getRelationalExprValues(valuesMap);
```
Comment (krdev) [inline json-path/src/main/java/com/jayway/jsonpath/internal/path/CompiledPath.java]: try to use our coding standards like @Nonnull , final for variables that are final and not change

File: json-path/src/main/java/com/jayway/jsonpath/internal/path/CompiledPath.java [hunk 6]
```diff
@@ -142,4 +147,17 @@ public boolean isFunctionPath() {
...
+    
+    /**
+     * Fetches the relational expression values from this path.
+     *
+     * @return EvaluationContext containing results of evaluation
+     */
+    @SuppressWarnings("unchecked")
+	@Override
+    public ArrayList<SimpleEntry<String, String>> getRelationalExprValues(){
+    	ArrayList<SimpleEntry<String, String>> valuesMap = new ArrayList<SimpleEntry<String, String>>();
```
Comment (krdev) [inline json-path/src/main/java/com/jayway/jsonpath/internal/path/CompiledPath.java]: make this final

File: json-path/src/main/java/com/jayway/jsonpath/Predicate.java (hunk 3 above)
Comment (krdev) [inline json-path/src/main/java/com/jayway/jsonpath/Predicate.java]: Change all interfaces to take and return List<SimpleEntry<String, String>> - the concrete implement, the one creating the list, will be the one that should decide to use a ArrayList

File: json-path/src/test/java/com/jayway/jsonpath/old/FilterTest.java [hunk 7]
```diff
@@ -348,6 +350,12 @@ public boolean apply(PredicateContext ctx) {
                 }
                 return false;
             }
+
+			@Override
+			public void getRelationalExprValues(ArrayList<SimpleEntry<String, String>> valuesMap) {
+				// TODO Auto-generated method stub
+				
+			}
```
Comment (krdev) [inline json-path/src/test/java/com/jayway/jsonpath/old/FilterTest.java]: Can you add a couple of tests for the filter operators listed in https://github.com/json-path/JsonPath ? make sure you do in, nin, subset-of, size, empty.  You can add a couple for relational operators don't have to cover all.

**Review by krdev — COMMENTED** [review]
ping me when you have completed the review.

File: json-path/src/main/java/com/jayway/jsonpath/internal/JsonContext.java (hunk 4 above)
Comment (ravitejakula) [reply json-path/src/main/java/com/jayway/jsonpath/internal/JsonContext.java]: Done.

File: json-path/src/main/java/com/jayway/jsonpath/ReadContext.java (hunk 1 above)
Comment (ravitejakula) [reply json-path/src/main/java/com/jayway/jsonpath/ReadContext.java]: Done

File: json-path/src/main/java/com/jayway/jsonpath/Predicate.java (hunk 3 above)
Comment (ravitejakula) [reply json-path/src/main/java/com/jayway/jsonpath/Predicate.java]: Done

Comment (ravitejakula) [reply json-path/src/main/java/com/jayway/jsonpath/Predicate.java]: Done

File: json-path/src/main/java/com/jayway/jsonpath/internal/JsonContext.java (hunk 2 above)
Comment (ravitejakula) [reply json-path/src/main/java/com/jayway/jsonpath/internal/JsonContext.java:149]: To make sure we are not returning null map when, processing a jsonpath using readroot, throws an exception.

File: json-path/src/main/java/com/jayway/jsonpath/internal/path/CompiledPath.java (hunk 6 above)
Comment (ravitejakula) [reply json-path/src/main/java/com/jayway/jsonpath/internal/path/CompiledPath.java]: Done

File: json-path/src/main/java/com/jayway/jsonpath/internal/path/CompiledPath.java (hunk 5 above)
Comment (ravitejakula) [reply json-path/src/main/java/com/jayway/jsonpath/internal/path/CompiledPath.java]: Done

File: json-path/src/test/java/com/jayway/jsonpath/old/FilterTest.java (hunk 7 above)
Comment (ravitejakula) [reply json-path/src/test/java/com/jayway/jsonpath/old/FilterTest.java]: Done.

**Pull request merged**